internal/app/upstream/client.go

```diff
@@ -55,11 +79,206 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	// TODO: configure grpc options.
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+func (m *client) OpenStream(ctx context.Context, request *v2.DiscoveryRequest) (<-chan *Response, error) {
+	response := make(chan *Response, 1)
+	if err := validateResourceType(request); err != nil {
+		close(response)
+		return nil, err
+	}
+
+	continueChan := make(chan bool, 1)
+	continueChan <- true
+	go func() {
+		for {
+			select {
+			case <-continueChan:
```

Comment:
@jessicayuen This is the alternate approach we discussed about. The error is not sent back to the caller and retried continuously. There might be other ways to achieve the same, but this approach is as good as others.